add channels last support for thnn_conv2d (non-dilated)

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61338 add channels last support for FractionalMaxPool2d on CPU
* #58348 opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU
* #55864 enable BFloat16 mkldnn_convolution on both contiguous and channels last memory format
* #55584 add channels last (2d) support for mkldnn_convolution
* #51185 add channels last support for ConvTranspose2d
* #50573 add channels last support for PixelShuffle and PixelUnshuffle
* #50247 add channels last support for ChannelShuffle
* #49984 add channel last support for MaxUnpool2d
* #49821 add channels last for GroupNorm
* **#49582 add channels last support for thnn_conv2d (non-dilated)**

Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

mingfeima committed:
add channels last support for thnn_conv2d (non-dilated)

[ghstack-poisoned]

mingfeima commented:
#### Initiative of this patch
Due to performance or functionality reasons, `nn.Conv2d` will choose **mkldnn_convolution** or **thnn_conv2d** on CPU depending on input tensor size and kernel size, refer to [use_mkldnn definition](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L227). So in order to claim that `nn.Conv2d` supports channels last memory format on CPU, need to make sure the underlying implementations support CL format. **mkldnn_convolution** CL support will be covered in later patches, this PR aims at providing CL support for **thnn_conv2d**.

#### Performance
Measured Conv2d test configs #40610 on single core of Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz, mkldnn path is manually disabled with `torch._C._set_mkldnn_enabled(False)` so as to test **thnn_conv2d**:

Notice that on grouped conv CL performs worse than Contiguous in the following table, but grouped conv will end up in mkldnn with default setting (mkldnn has much better perf with group conv than the gemm based algorithm)

input size | output channels | kernel | stride | padding | groups | before (nchw) | after (nchw) | after (nhwc)
-- | -- | -- | -- | -- | -- | -- | -- | --
[1,3,224,224] | 64 | 7 | 2 | 3 | 1 | 4.443 | 4.435 | 4.488
[1,64,56,56] | 128 | 1 | 1 | 0 | 1 | 0.337 | 0.337 | 0.360
[1,128,56,56] | 128 | 3 | 1 | 1 | 32 | 1.784 | 1.798 | 9.012
[1,128,56,56] | 256 | 1 | 1 | 0 | 1 | 1.117 | 1.116 | 1.155
[1,64,56,56] | 256 | 1 | 1 | 0 | 1 | 0.773 | 0.774 | 0.783
[1,256,56,56] | 128 | 1 | 1 | 0 | 1 | 0.994 | 1.006 | 1.084
[1,128,56,56] | 128 | 3 | 1 | 1 | 32 | 1.787 | 1.784 | 9.013
[1,128,56,56] | 256 | 1 | 1 | 0 | 1 | 1.117 | 1.117 | 1.159
[1,256,56,56] | 128 | 1 | 1 | 0 | 1 | 0.994 | 0.995 | 1.085
[1,128,56,56] | 128 | 3 | 1 | 1 | 32 | 1.783 | 1.786 | 9.004
[1,128,56,56] | 256 | 1 | 1 | 0 | 1 | 1.123 | 1.115 | 1.157
[1,256,56,56] | 256 | 1 | 1 | 0 | 1 | 2.057 | 2.059 | 2.136
[1,256,56,56] | 256 | 3 | 2 | 1 | 32 | 4.083 | 4.101 | 4.186
[1,256,28,28] | 512 | 1 | 1 | 0 | 1 | 1.059 | 1.057 | 1.073
[1,256,56,56] | 512 | 1 | 2 | 0 | 1 | 1.395 | 1.400 | 1.149
[1,512,28,28] | 256 | 1 | 1 | 0 | 1 | 1.048 | 1.050 | 1.109
[1,256,28,28] | 256 | 3 | 1 | 1 | 32 | 1.379 | 1.383 | 2.936
[1,256,28,28] | 512 | 1 | 1 | 0 | 1 | 1.057 | 1.056 | 1.072
[1,512,28,28] | 256 | 1 | 1 | 0 | 1 | 1.048 | 1.051 | 1.106
[1,256,28,28] | 256 | 3 | 1 | 1 | 32 | 1.376 | 1.381 | 2.942
[1,256,28,28] | 512 | 1 | 1 | 0 | 1 | 1.059 | 1.065 | 1.073
[1,512,28,28] | 256 | 1 | 1 | 0 | 1 | 1.049 | 1.056 | 1.110
[1,256,28,28] | 256 | 3 | 1 | 1 | 32 | 1.377 | 1.387 | 2.933
[1,256,28,28] | 512 | 1 | 1 | 0 | 1 | 1.058 | 1.056 | 1.072
[1,512,28,28] | 512 | 1 | 1 | 0 | 1 | 2.014 | 2.008 | 2.059
[1,512,28,28] | 512 | 3 | 2 | 1 | 32 | 2.263 | 2.284 | 1.583
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.200 | 1.199 | 1.136
[1,512,28,28] | 1024 | 1 | 2 | 0 | 1 | 1.356 | 1.359 | 1.175
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.239 | 1.235 | 1.164
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.380 | 1.395 | 1.301
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.192 | 1.190 | 1.140
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.234 | 1.235 | 1.163
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.375 | 1.394 | 1.303
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.192 | 1.195 | 1.143
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.240 | 1.235 | 1.165
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.384 | 1.394 | 1.309
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.195 | 1.189 | 1.134
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.234 | 1.231 | 1.164
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.380 | 1.399 | 1.301
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.196 | 1.188 | 1.141
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.234 | 1.237 | 1.168
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.379 | 1.388 | 1.303
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.192 | 1.192 | 1.140
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.234 | 1.234 | 1.166
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.381 | 1.389 | 1.302
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.189 | 1.194 | 1.138
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.235 | 1.235 | 1.161
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.379 | 1.397 | 1.312
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.198 | 1.191 | 1.143
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.236 | 1.233 | 1.163
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.379 | 1.391 | 1.306
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.191 | 1.190 | 1.137
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.236 | 1.232 | 1.168
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.375 | 1.389 | 1.303
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.194 | 1.195 | 1.136
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.243 | 1.241 | 1.161
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.379 | 1.398 | 1.305
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.195 | 1.189 | 1.141
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.237 | 1.232 | 1.163
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.379 | 1.391 | 1.304
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.193 | 1.188 | 1.136
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.237 | 1.231 | 1.166
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.380 | 1.397 | 1.305
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.191 | 1.193 | 1.136
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.233 | 1.234 | 1.163
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.377 | 1.399 | 1.302
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.194 | 1.191 | 1.140
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.236 | 1.237 | 1.162
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.376 | 1.391 | 1.305
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.194 | 1.194 | 1.144
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.233 | 1.232 | 1.169
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.381 | 1.397 | 1.300
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.191 | 1.189 | 1.140
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.236 | 1.235 | 1.164
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.379 | 1.397 | 1.306
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.199 | 1.187 | 1.136
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.239 | 1.241 | 1.162
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.377 | 1.404 | 1.305
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.192 | 1.189 | 1.136
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.235 | 1.233 | 1.166
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.379 | 1.407 | 1.310
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.202 | 1.193 | 1.151
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.239 | 1.234 | 1.163
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.376 | 1.393 | 1.305
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.194 | 1.199 | 1.134
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.234 | 1.236 | 1.162
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.379 | 1.395 | 1.305
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.191 | 1.194 | 1.138
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.235 | 1.234 | 1.164
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.377 | 1.398 | 1.301
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.192 | 1.191 | 1.137
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.232 | 1.233 | 1.168
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 1.380 | 1.394 | 1.310
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.191 | 1.189 | 1.135
[1,1024,14,14] | 1024 | 1 | 1 | 0 | 1 | 2.336 | 2.333 | 2.280
[1,1024,14,14] | 1024 | 3 | 2 | 1 | 32 | 1.711 | 1.735 | 1.288
[1,1024,7,7] | 2048 | 1 | 1 | 0 | 1 | 1.620 | 1.615 | 1.746
[1,1024,14,14] | 2048 | 1 | 2 | 0 | 1 | 1.720 | 1.728 | 1.763
[1,2048,7,7] | 1024 | 1 | 1 | 0 | 1 | 1.630 | 1.629 | 1.763
[1,1024,7,7] | 1024 | 3 | 1 | 1 | 32 | 1.492 | 1.503 | 1.166
[1,1024,7,7] | 2048 | 1 | 1 | 0 | 1 | 1.620 | 1.614 | 1.749
[1,2048,7,7] | 1024 | 1 | 1 | 0 | 1 | 1.624 | 1.640 | 1.760
[1,1024,7,7] | 1024 | 3 | 1 | 1 | 32 | 1.489 | 1.510 | 1.172
[1,1024,7,7] | 2048 | 1 | 1 | 0 | 1 | 1.630 | 1.616 | 1.750

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"




[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"




[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"




[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"




[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"




[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"




[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"




[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

glaringlee commented on aten/src/ATen/native/cpu/Unfold2d.cpp:
```diff
@@ -255,6 +354,68 @@ static void unfolded2d_copy(
+    int64_t input_width,
+    int64_t output_height,
+    int64_t output_width) {
+  // algorithm on channels last is dual problem to contiguous format...
```
Just curious on this comment. On what aspects, you define this algorithm as a dual problem to the contiguous format?

mingfeima replied:
The output shape of Unfold (or im2col) in ChannelsLast format is transpose of Contiguous format.

Suppose `batch_size=1` to simply:
```bash
### contiguous (nchw), tensor shape in 2d view:
input: [input_channel, input_height * input width]
column: [input_channel * kernel_height * kernel_width, output_height * output_width]
weight: [output_channel, input_channel * kernel_height * kernel_width]
output = weight * column

### channels last (nhwc), tensor shape in 2d view:
input: [input_height * input_width, input_channel]
column: [output_height * output_width, kernel_height * kernel_width * input_channel]
weight: [output_channel, kernel_height * kernel_width * input_channel]
output = column * weight.transpose()
```

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007050).

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007050).

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima commented:
rebased!

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007050).

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for thnn_conv2d (non-dilated)"


Differential Revision: [D26007050](https://our.internmc.facebook.com/intern/diff/D26007050)

[ghstack-poisoned]

mingfeima commented:
rebased!

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007050).

VitalyFedyunin commented:
Reverting as it might be offender to Iphone perf regression, please do nothing at thing point, might be false positive, I'm investigating.

mingfeima commented:
@VitalyFedyunin  The current code base in `ConvolutionMM2d.cpp` has replaced `.mm()` with `cpublas::gemm` which shall be able to save several dispatch process.

I will rewrite this one accordingly, and collect new performance data. @VitalyFedyunin
